### PR TITLE
Implement Navigate To service

### PR DESCRIPTION
**Pull request by vasily-kirichenko**
image

### CODE REVIEW

File: vsintegration/src/FSharp.Editor/NavigateToSearchService.fs [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+
+[<ExportLanguageService(typeof<INavigateToSearchService>, FSharpCommonConstants.FSharpLanguageName); Shared>]
+type internal FSharpNavigateToSearchService 
+    [<ImportingConstructor>] 
+    (
+        checkerProvider: FSharpCheckerProvider,
+        projectInfoManager: ProjectInfoManager
+    ) =
+
+    let itemsByDocumentId = ConcurrentDictionary<DocumentId, (int * NavigableItem [])>()
```
Comment (dsyme) [inline vsintegration/src/FSharp.Editor/NavigateToSearchService.fs]: Should this be a ConditionalWeakTable?

**Comment by dsyme** [comment]
Wow, fantastic!

**Comment by dsyme** [comment]
This looks fantastic :)

**Comment by vasily-kirichenko** [comment]
@dsyme :) Done.

**Comment by cartermp** [comment]
Looks awesome!

Is the Navigate To box in the center due to your own settings?

**Comment by vasily-kirichenko** [comment]
@cartermp yes :)

**Comment by cartermp** [comment]
I've been playing with this today and noticed two main things:

* Fuzzy matching doesn't seem to work (e.g., typing "wb" in the FSharpEx project won't show the `WriterBuilder` type in the result set)
* It seems to drop certain symbols after a while.  I was able to search for types just fine, but after I closed by laptop and opened it again a few minutes later (didn't open/close VS), I was unable to search for any types.

**Comment by vasily-kirichenko** [comment]
Hm, I was not aware, there is such a thing as fuzzy matching. 

The second problem looks mysterious :(

**Comment by vasily-kirichenko** [comment]
I don't think VS supports fuzzy matching at all:

image

**Comment by vasily-kirichenko** [comment]
Ah, if you use upper case, then it kind of works:

image

However, I'm not sure I understand why `IsActive` is in the results, I mean what was happened to the `C` character?

**Comment by vasily-kirichenko** [comment]
C#/VB use `PatternMatcher` http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/PatternMatching/PatternMatcher.cs,22 I think we should use it in conjunction with our existing logic, because `PatternMatcher` is unlikely to support searching for operators and double quoted identifiers, for example (however, I may be wrong).

The problem I cannot find it in any of NuGet packages we use. It seems it should be in `Microsoft.CodeAnalysis.Workspaces.Common`, but it's not. There is no such a package as `Microsoft.CodeAnalysis.Workspaces` either.

Thoughts?

**Comment by cartermp** [comment]
@CyrusNajmabadi, any thoughts to address @vasily-kirichenko's investigation for fuzzy matching in NavigateTo?

**Comment by cartermp** [comment]
As for the second issue I found, I think we shouldn't worry about it for now.  If I can reproduce it I'll report back.

**Comment by CyrusNajmabadi** [comment]
> However, I'm not sure I understand why IsActive is in the results, I mean what was happened to the C character?

So, let's clarify the terminology that Roslyn uses for 'matching'.  We currently use:

1. 'CamelCase' to mean the process of doing something like    ComProv in order to match CompletionProvider.  "CamelCase' is inaccurate and it would be better termed 'PascalCase'.   But CamelCase reflects the 'humpy' nature of the match.
2. 'Fuzzy', to reflect some degree of 'error tolerance'.  Meaning you can mispell a few characters, and we still find matches.

So, when you typed ICActive, our 'fuzzy' matcher said: that's closed enough to IsActive to count as a hit.  You may have just flubbed typing into the box.

**Comment by CyrusNajmabadi** [comment]
> The problem I cannot find it in any of NuGet packages we use. It seems it should be in Microsoft.CodeAnalysis.Workspaces.Common, but it's not. There is no such a package as Microsoft.CodeAnalysis.Workspaces either.

@Pilchie @jasonmal PatternMatcher exists at the workspace layer.  What's the best way for F# to be able to use it?

**Comment by CyrusNajmabadi** [comment]
> I think we should use it in conjunction with our existing logic, because PatternMatcher is unlikely to support searching for operators and double quoted identifiers, for example (however, I may be wrong).

You are correct.  The PatternMatcher is not really a general purpose piece of code.  It's tuned to the sorts of things people search for in C# and VB.  That shoudl work for F# for lots of cases (since you're very .net'ish), but you will likely still need to augment it with your own special language cases.

**Comment by vasily-kirichenko** [comment]
After merging with master I got this error:

image

Looks like `xxx.CodeAnalysis.yyy` are newer than beta-6 we are using:

image

**Comment by vasily-kirichenko** [comment]
Even RC packages has older version than VS RC2 uses:

image

**Comment by vasily-kirichenko** [comment]
Yes, updating packages to RC have not helped.

**Comment by KevinRansom** [comment]
@vasily-kirichenko 
I have resolved collisions and updated the references to rc2, is this ready to go?

@dotnet-bot test this please

**Comment by vasily-kirichenko** [comment]
@KevinRansom yes, it's ready. It does not use Roslyn's PatternMatcher to perform CamelCaps searching since it's not available via nuget packages yet. We should fix this later.

**Comment by vasily-kirichenko** [comment]
> PatternMatcher exists at the workspace layer. What's the best way for F# to be able to use it?

@Pilchie @jasonmal any updates on this? We would love to use `PatternMatcher` to make the "CamelCase" searching.

**Comment by forki** [comment]
image

**Comment by CyrusNajmabadi** [comment]
Honestly, right now the best thing to do would be to copy the code :)

We have started discussions with the Editor team to get the PatternMatcher cleaned up and available as a core VS service.  But that is certainly not happening for VS2017 RTM.  So, if you want to unblock yourself now, I would just copy our code.  Otherwise, we can try to push for this soon after RTM.

**Comment by vasily-kirichenko** [comment]
@CyrusNajmabadi I tried to port the code to F#, but it turned out there are quite a lot of it, so I gave up.

**Comment by CyrusNajmabadi** [comment]
image

**Comment by CyrusNajmabadi** [comment]
I'm surprised.  It shoudln't be that involved to do a nasty copy/paste job :)

**Comment by vasily-kirichenko** [comment]
@CyrusNajmabadi that's not copy-paste, that's porting to F#. And `PatternMatcher` depends on more code, that depends on even more and so on.

**Comment by CyrusNajmabadi** [comment]
right... but i think that set is reasonably tiny...

**Comment by CyrusNajmabadi** [comment]
Ok. Looking at it, i can't really see the problem.  I did notice that PatternMatcher pulls in WordSimilarityChecker in order to support "fuzzy matching" (which is distinct from CamelCaseMatching).  If you want, you can just literally cut out that functionality to pull in less.

**Comment by vasily-kirichenko** [comment]
Ok, I'll try again cutting out the fuzzy matcher.

**Comment by smoothdeveloper** [comment]
@vasily-kirichenko what about stuffing the C# code in a separate project (or existing c# project in the solution)?

**Comment by forki** [comment]
We would probably ilrepack or handle deployment,...

**Comment by CyrusNajmabadi** [comment]
I did hte exercise here: https://github.com/CyrusNajmabadi/PatternMatching

**Comment by CyrusNajmabadi** [comment]
It was definitely a fair amount of code to copy (if you want fuzzy matching).  If you don't, then it get's much smaller.

**Pull request merged**